[ParseableInterfaces] Support inheriting default arguments in module interfaces via '= super'

nathawes opened the pull request:
Since we try not to synthesize inherited initializers, we want to be explicit and print them in subclasses. Since the default argument kind was `Inherited`, though, we were printing them as:

```swift
override public init(value: Swift.Int = super)
```

which didn't parse. Harlan initially tried fixing this by changing how we print the module interfaces (https://github.com/apple/swift/pull/23933) but that approach isn't always correct. This PR changes how we consume them by updating the parser to accept the `= super` syntax in module interface files to mean the argument kind is `Inherited`.

I also added diagnostics for `= super` showing up where it shouldn't, but maybe that's overkill since module interfaces are generated?

Resolves rdar://49789274

Commit:
Add an attribute to inhert a param's default value from the overridden decl in module interfaces

Commit:
Add diagnostics for the new @_inheritedDefaultValue attribute

Commit:
Change from using the @_inheritedDefaultValue attribute to mark parameters with an inherited default argument to using '= super'

Commit:
Update tests from @_inheritedDefaultValue to '= super'

Commit:
Add test for the SwiftSyntax tree produced when parsing '= super' in a module interface

nathawes commented:
@swift-ci please test

harlanhaskins commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -66,10 +66,22 @@ void Parser::DefaultArgumentInfo::setFunctionContext(
+  // Handle the special '= super' syntax for inherited default values in module
+  // interface files
+  if (P.SF.Kind == SourceFileKind::Interface) {
+    if (P.Tok.is(tok::kw_super) && P.peekToken().isAny(tok::comma, tok::r_paren)) {
```
Can you expand this comment to explain that you're explicitly looking for `super,` and `super)`, and that future syntactic expansion around inherited default args would break this?

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1281,6 +1281,16 @@ WARNING(expr_dynamic_lookup_swift3_objc_inference,none,
+      "parameter of the overridden initializer has a default value", ())
+ERROR(inherited_default_value_used_in_non_overriding_constructor,none,
+      "default value inheritance via 'super' can only be used on the "
+      "parameters of overriding initializers", ())
```
What happens for inherited convenience initializers (which are only sort of overriding)?

slavapestov replied:
Inherited convenience inits are visible via name lookup like other inherited decls when the right flag on the ClassDecl is set. Only designated inits get explicitly synthesized overrides.

nathawes replied:
I've updated the checks and wording of the diagnostics to restrict this to designated inits now.

harlanhaskins reviewed (approved):
Thanks for tackling this! And thanks for adding Syntax tests too.

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1737,10 +1737,74 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
+
+  auto *DC = PD->getInnermostDeclContext();
+  const SourceFile *SF = DC->getParentSourceFile();
+  if (!SF || SF->Kind != SourceFileKind::Interface)
```
I'd suggest *asserting* this, not just bailing out early. If the syntax makes it through the parser in the future, we want to know what needs to change.

nathawes replied:
Done!

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1737,10 +1737,74 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
+      TC.diagnose(PD->getLoc(),
+                  diag::corresponding_param_not_defaulted);
+    }
+    // We found a match: it's valid
```
…or we diagnosed, yeah?

nathawes replied:
Oops, yeah. I reworked this section a bit, because I realized i didn't need the loop through the chain of overridden decls when the corresponding default is inherited too as that default be checked separately anyway.

jrose-apple commented on test/ParseableInterface/default-args.swift:
```diff
@@ -3,8 +3,29 @@
+  // CHECK: init(x: Int = 3)
+	public init(x: Int = 3) {}
+	// CHECK: foo(y: Int = 42)
+	public func foo(y: Int = 42) {}
```
Mix of tabs and spaces in this file.

nathawes replied:
Thanks – fixed!

jrose-apple commented on test/ParseableInterface/inherited-defaults-checks.swiftinterface:
```diff
@@ -0,0 +1,38 @@
+// swift-module-flags: -module-name InheritedDefaults
+
+// RUN: %empty-directory(%t)
+// RUN: %target-typecheck-verify-swift
```
What's this extra step buying us?

nathawes replied:
Oops – only meant to keep the first line.

jrose-apple commented on test/ParseableInterface/inherited-defaults-checks.swiftinterface:
```diff
@@ -0,0 +1,38 @@
+
+public class Indirect: Direct {
+  public override init(x: Int = super, y: Int)
+  public override init(other: Int = super, value: Int)
```
Nitpick: none of your test cases have the `super` default on any parameter but the first.

nathawes replied:
Updated the tests to vary this a little more.

jrose-apple reviewed (commented):
Looks pretty good, only minor comments!

Commit:
Simplify diagnostics for inherited default argument values in module interfaces

Also:
- additionally require the containing and overridden initializers are
  designated, as that's the only case in which we should produce the '= super'
  syntax in module interfaces
- Add notes to point out the locations of the overriden initializer when it's
  not designated, and the corresponding parameter in that initializer when it
  doesn't have a default argument to inherit.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test Linux

jrose-apple reviewed (commented):
Do you think we need a SILGen or execution test to make sure the value's getting parsed *correctly?* That can probably be a follow-up commit, but it seems like a good idea to me.

nathawes commented on test/ParseableInterface/default-args.swift:
```diff
@@ -3,8 +3,34 @@
+  }
+
+  // CHECK-NOT: init(convInit: Int = super)
+  // CHECK: override {{(public )?}}init(x: Int = super)
```
It's just `override` in the `swift-ide-test -print-module` output versus `override public` in the module interface file. The `swift-ide-test -print-module` output doesn't show `public` anywhere.

jrose-apple replied:
Ah, okay. I guess that's all right.

nathawes commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1737,10 +1737,60 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
+        PD->getLoc(),
+        diag::inherited_default_value_used_in_non_overriding_constructor);
+    if (overridden)
+      TC.diagnose(overridden->getLoc(), diag::overridden_here);
```
I didn't know that overload existed – thanks!

nathawes replied:
Done.

Commit:
Check '= super' is actually parsed correctly in module interfaces via an execution test

Also pass the decls themselves rather than their locations when diagnosing
incorrect usage of '= super'.

nathawes commented:
@swift-ci please test

nathawes commented:
@jrose-apple I've added the execution test, but I haven't actually written one before. Does this look ok?

jrose-apple reviewed (approved):
Thank you!

jrose-apple commented on test/ParseableInterface/inherited-defaults-execution.swift:
```diff
@@ -0,0 +1,47 @@
+//
+// RUN: %target-build-swift -I%t -L%t -lInherited -o %t/main %target-rpath(%t) %t/main.swift -swift-version 5
+// RUN: %target-codesign %t/main %t/%target-library-name(Inherited)
+// RUN: %target-run %t/main | %FileCheck --check-prefix=OUTPUT %s
```
Oops, look like this broke device testing. In order to make sure libInherited.dylib gets copied to the device, you have to pass it as an argument to `%t/main`. (That's the hack we came up with, anyway: pretend it's an input.)

jrose-apple replied:
Oh, you're way ahead of me. :-)

nathawes replied:
Yeah, I wasn't sure what I was missing, but @harlanhaskins helped me out!
